google-cloud-logging/src/main/java/com/google/cloud/logging/Instrumentation.java

```diff
@@ -0,0 +1,196 @@
+                        .build()))
+            .build();
+    LogEntry entry =
+        LogEntry.of(
+            JsonPayload.of(
+                Struct.newBuilder()
+                    .putAllFields(
+                        ImmutableMap.of(
+                            DIAGNOSTIC_INFO_KEY,
+                            Value.newBuilder().setStructValue(instrumentation).build()))
+                    .build()));
+    return entry;
+  }
+
+  private static ListValue generateLibrariesList(
+      String libraryName, String libraryVersion, ListValue existingLibraryList) {
+    if (Strings.isNullOrEmpty(libraryName) || !libraryName.startsWith(JAVA_LIBRARY_NAME_PREFIX))
+      libraryName = JAVA_LIBRARY_NAME_PREFIX;
+    if (Strings.isNullOrEmpty(libraryVersion)) {
+      libraryVersion = getLibraryVersion(Instrumentation.class.getClass());
+    }
+    Struct libraryInfo = createInfoStruct(libraryName, libraryVersion);
+    ListValue.Builder libraryList = ListValue.newBuilder();
+    // Append first the library info for this library
+    libraryList.addValues(Value.newBuilder().setStructValue(libraryInfo).build());
+    if (existingLibraryList != null) {
+      for (Value val : existingLibraryList.getValuesList()) {
+        if (val.hasStructValue()) {
+          try {
+            String name =
+                val.getStructValue().getFieldsOrThrow(INSTRUMENTATION_NAME_KEY).getStringValue();
+            if (Strings.isNullOrEmpty(name) || !name.startsWith(JAVA_LIBRARY_NAME_PREFIX)) continue;
+            String version =
+                val.getStructValue().getFieldsOrThrow(INSTRUMENTATION_VERSION_KEY).getStringValue();
+            if (Strings.isNullOrEmpty(version)) continue;
+            libraryList.addValues(
+                Value.newBuilder().setStructValue(createInfoStruct(name, version)).build());
+            if (libraryList.getValuesCount() == MAX_DIAGNOSTIC_ENTIES) break;
+          } catch (Exception ex) {
+            System.err.println("ERROR: unexpected exception in generateLibrariesList: " + ex);
+          }
+        }
+      }
+    }
+    return libraryList.build();
+  }
+
+  private static Struct createInfoStruct(String libraryName, String libraryVersion) {
+    return Struct.newBuilder()
+        .putAllFields(
+            ImmutableMap.of(
+                INSTRUMENTATION_NAME_KEY,
+                    Value.newBuilder().setStringValue(truncateValue(libraryName)).build(),
+                INSTRUMENTATION_VERSION_KEY,
+                    Value.newBuilder().setStringValue(truncateValue(libraryVersion)).build()))
+        .build();
+  }
+
+  /**
+   * The helper method used to set a status of a flag which indicates if instrumentation info
+   * already written or not.
+   *
+   * @param value {boolean} The value to be set.
+   * @returns The value of the flag before it is set.
+   */
+  public static boolean setInstrumentationStatus(boolean value) {
```

Comment:
nit: Use of two methods seems unnecessary complex and the naming can be confusing to a reader. It is possible to use a synchronized block of code with some object:
```java
synchronized(mutexObject) {
  // your code here
}
```
The object can be a variable of type `Boolean` (but not `boolean` value).

Reply:
Sure - added a separate `Object` to be used for synchronization